[dtensor] add StrategyType and TupleStrategy

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99374
* #99489
* #99436
* #99373
* __->__ #99435
* #99062
* #98452

This PR refactors the current StrategyList. It introduces a
StrategyType, which is the base class of Strategy, and it have
two sub strategies:

1. Refactor the previous StrategyList to OpStrategy
2. Add TupleStrategy, the new strategy added to deal with tuple cases where
it could return multiple different OpStrategy for an op.

This would help support a more complicated op and unblocks compile mode
FSDP

wanchaol committed:
[dtensor] add StrategyType and TupleStrategy

This PR refactors the current StrategyList. It introduces a
StrategyType, which is the base class of Strategy, and it have
two sub strategies:

1. Refactor the previous StrategyList to OpStrategy
2. Add TupleStrategy, the new strategy added to deal with tuple cases where
it could return multiple different OpStrategy for an op.

This would help support a more complicated op and unblocks compile mode
FSDP

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] add StrategyType and TupleStrategy"


This PR refactors the current StrategyList. It introduces a
StrategyType, which is the base class of Strategy, and it have
two sub strategies:

1. Refactor the previous StrategyList to OpStrategy
2. Add TupleStrategy, the new strategy added to deal with tuple cases where
it could return multiple different OpStrategy for an op.

This would help support a more complicated op and unblocks compile mode
FSDP

[ghstack-poisoned]

mrshenli commented on torch/distributed/_tensor/op_schema.py:
```diff
@@ -54,17 +54,49 @@ def __str__(self) -> str:
+class TupleStrategy(StrategyType):
+    """
+    TupleStrategy represents the output strategy of this op is a tuple
+    of strategy lists, this would be used if the output strategies might
```
do we always return a TupleStrategy if the output is a tuple of tensors? Or do we only use this when those tensors have different strategies?

wanchaol replied:
This is a good point, let me think through this a bit and see if we should consolidate on always returning TupleStrategy

wanchaol replied:
for `List[Tensor]` return I think we should have a single OpStrategy, where each elements share the same output spec, but for `Tuple[Tensor, ..]` I think we should return TupleStrategy regardless of whether the strategies are same or not, I'll update the comment.

wanchaol committed:
Update on "[dtensor] add StrategyType and TupleStrategy"


This PR refactors the current StrategyList. It introduces a
StrategyType, which is the base class of Strategy, and it have
two sub strategies:

1. Refactor the previous StrategyList to OpStrategy
2. Add TupleStrategy, the new strategy added to deal with tuple cases where
it could return multiple different OpStrategy for an op.

This would help support a more complicated op and unblocks compile mode
FSDP

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] add StrategyType and TupleStrategy"


This PR refactors the current StrategyList. It introduces a
StrategyType, which is the base class of Strategy, and it have
two sub strategies:

1. Refactor the previous StrategyList to OpStrategy
2. Add TupleStrategy, the new strategy added to deal with tuple cases where
it could return multiple different OpStrategy for an op.

This would help support a more complicated op and unblocks compile mode
FSDP

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] add StrategyType and TupleStrategy"


This PR refactors the current StrategyList. It introduces a
StrategyType, which is the base class of Strategy, and it have
two sub strategies:

1. Refactor the previous StrategyList to OpStrategy
2. Add TupleStrategy, the new strategy added to deal with tuple cases where
it could return multiple different OpStrategy for an op.

This would help support a more complicated op and unblocks compile mode
FSDP

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] add StrategyType and TupleStrategy"


This PR refactors the current StrategyList. It introduces a
StrategyType, which is the base class of Strategy, and it have
two sub strategies:

1. Refactor the previous StrategyList to OpStrategy
2. Add TupleStrategy, the new strategy added to deal with tuple cases where
it could return multiple different OpStrategy for an op.

This would help support a more complicated op and unblocks compile mode
FSDP

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] add StrategyType and TupleStrategy"


This PR refactors the current StrategyList. It introduces a
StrategyType, which is the base class of Strategy, and it have
two sub strategies:

1. Refactor the previous StrategyList to OpStrategy
2. Add TupleStrategy, the new strategy added to deal with tuple cases where
it could return multiple different OpStrategy for an op.

This would help support a more complicated op and unblocks compile mode
FSDP

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] add StrategyType and TupleStrategy"


This PR refactors the current StrategyList. It introduces a
StrategyType, which is the base class of Strategy, and it have
two sub strategies:

1. Refactor the previous StrategyList to OpStrategy
2. Add TupleStrategy, the new strategy added to deal with tuple cases where
it could return multiple different OpStrategy for an op.

This would help support a more complicated op and unblocks compile mode
FSDP

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] add StrategyType and TupleStrategy"


This PR refactors the current StrategyList. It introduces a
StrategyType, which is the base class of Strategy, and it have
two sub strategies:

1. Refactor the previous StrategyList to OpStrategy
2. Add TupleStrategy, the new strategy added to deal with tuple cases where
it could return multiple different OpStrategy for an op.

This would help support a more complicated op and unblocks compile mode
FSDP

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] add StrategyType and TupleStrategy"


This PR refactors the current StrategyList. It introduces a
StrategyType, which is the base class of Strategy, and it have
two sub strategies:

1. Refactor the previous StrategyList to OpStrategy
2. Add TupleStrategy, the new strategy added to deal with tuple cases where
it could return multiple different OpStrategy for an op.

This would help support a more complicated op and unblocks compile mode
FSDP

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] add StrategyType and TupleStrategy"


This PR refactors the current StrategyList. It introduces a
StrategyType, which is the base class of Strategy, and it have
two sub strategies:

1. Refactor the previous StrategyList to OpStrategy
2. Add TupleStrategy, the new strategy added to deal with tuple cases where
it could return multiple different OpStrategy for an op.

This would help support a more complicated op and unblocks compile mode
FSDP

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] add StrategyType and TupleStrategy"


This PR refactors the current StrategyList. It introduces a
StrategyType, which is the base class of Strategy, and it have
two sub strategies:

1. Refactor the previous StrategyList to OpStrategy
2. Add TupleStrategy, the new strategy added to deal with tuple cases where
it could return multiple different OpStrategy for an op.

This would help support a more complicated op and unblocks compile mode
FSDP

[ghstack-poisoned]
